docs/proposals/role_decomposition.md

```diff
@@ -0,0 +1,322 @@
+# Scaffolding for decomposing large roles
+
+## Why?
+
+Currently we have roles that are very large and encompass a lot of different
+components. This makes for a lot of logic required within the role, can
+create complex conditionals, and increases the learning curve for the role.
+
+## How?
+
+Creating a guide on how to approach breaking up a large role into smaller,
+component based, roles. Also describe how to develop new roles, to avoid creating
+large roles.
+
+## Proposal
+
+Create a new guide or append to the current contributing guide a process for
+identifying large roles that can be split up, and how to compose smaller roles
+going forward.
+
+### Large roles
+
+A role should be considered for decomposition if it:
+
+1) Configures/installs more than one product.
+1) Can configure multiple variations of the same product that can live
+side by side.
+1) Has different entry points for upgrading and installing a product
```

Comment:
I agree that this is the tricky part.  Knowing when and where to separate a role so that non-shared parts are abstracted is difficult.  That being said the ability to configure smaller pieces of a master (your example) IMHO, far outweigh the monolithic approach which requires logic, special variables, hard coded data, etc.  

> Given the current ansible has still some issues wrt. resource sharing (e.g can not include handlers from different roles properly)

Ansible has attempted to make this easier with `include_role` but you are correct that there are issues with it.
> it is non-trivial task to break install/upgrade role into two install and upgrade roles without increasing redundancy.

agreed.